### PR TITLE
Install query

### PR DESCRIPTION
**Pull request by apata**
### Changes

Please describe the changes made in the pull request here.

Below you'll find a checklist. For each item on the list, check one option and delete the other.

### Tests
- [ ] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [ ] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [ ] This change does not need a documentation update

### Dark mode
- [ ] The UI has been tested both in dark and light mode
- [ ] This PR does not change the UI

### CODE REVIEW

File: assets/js/dashboard/stats/modals/breakdown-modal.js [hunk 1]
```diff
@@ -84,24 +85,57 @@ export default function BreakdownModal({
...
+    getNextPageParam: (lastPageResults, _, lastPageIndex) => lastPageResults.length === LIMIT ? lastPageIndex + 1 : null,
+    initialPageParam: 1,
+    queryFn,
+  })
 
-  useMountedEffect(() => { debouncedFetchData() }, [search])
+  const queryClient = useQueryClient()
 
-  useMountedEffect(() => { fetchNextPage() }, [page])
+  useEffect(() => {
```
Comment (RobertJoonas) [inline assets/js/dashboard/stats/modals/breakdown-modal.js:126]: This `useEffect` here makes sure that we always cache only a single page of results.

If we imagine the following scenario:

1) user opens the modal
2) clicks load more 2 times
3) closes the modal
4) reopens the modal

Then,

* **without this `useEffect`** - after step 4) the user would instantly see the cached 3 pages of results. But since that data is stale, we'll still want to refresh all of it. That will take 3 API calls.
* **with this `useEffect`** - after step 4) we only get the first page of results from cache and make a single API call to refresh it.

So basically, it simply reduces load on the server.

File: assets/js/dashboard/stats/modals/breakdown-modal.js [hunk 2]
```diff
@@ -84,24 +85,57 @@ export default function BreakdownModal({
...
-  const [initialLoading, setInitialLoading] = useState(true)
-  const [loading, setLoading] = useState(true)
   const [search, setSearch] = useState('')
-  const [results, setResults] = useState([])
-  const [page, setPage] = useState(1)
-  const [moreResultsAvailable, setMoreResultsAvailable] = useState(false)
-  const searchBoxRef = useRef(null)
 
-  useEffect(() => { fetchData() }, [])
+  const queryFn = async ({ pageParam, queryKey }) => {
```
Comment (ukutaht) [inline assets/js/dashboard/stats/modals/breakdown-modal.js:95]: Given that the name `query` is already very overloaded in this codebase, I wonder if it would be possible to encapsulate the `react-query` stuff so that we can call it by a different name. Maybe a custom hook like `useApiClient`?

**Review by ukutaht — APPROVED** [review]
Looks good! I really like the UX improvement from caching. One comment about naming below.

File: assets/js/dashboard/stats/modals/breakdown-modal.js (hunk 2 above)
Comment (RobertJoonas) [reply assets/js/dashboard/stats/modals/breakdown-modal.js]: Agreed. I've created a new `useAPIClient` hook that also handles pagination. https://github.com/plausible/analytics/pull/4361/commits/0c53aec263af195b0ed77ff31100c579e8319fff.

Not every report needs pagination, so if we're planning to migrate other dashboard reports to use React Query, I was thinking we could create a new api hook, or make pagination optional somehow.

The Google Search Terms modal could nicely use the `useAPIClient` though, with built in pagination logic.

File: assets/js/dashboard/hooks/api-client.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+// * `getRequestParams` - a function that takes the `key` prop as an argument,
+//   and returns `[query, params]` which will be used by `queryFn` that actually
+//   calls the API.
+
+// ### Optional props
+
+// * `afterFetchData` - a function to call after data has been fetched.
+
+// * `afterFetchNextPage` - a function to call after the next page has been fetched.
```
Comment (apata) [inline assets/js/dashboard/hooks/api-client.js:27]: **Ultranitpick, non-blocking:** if you use [JSDoc code comment format](https://jsdoc.app/about-getting-started), the IDE shows it in a nicer way when you hover the function where it is used. 
```
/**
 * A wrapper for the React Query library. Constructs the necessary options
 * (including pagination config) to pass into the `useInfiniteQuery` hook.
 *
 * ### Required props
 *
 * @param {Array} key - The key under which the global "query" instance will live. Should
 *   be passed as a list of two elements - `[endpoint, { query }]`. The object
 *   can also contain additional values (such as `search`) to be used by:
 *   1) React Query, to determine the uniqueness of the query instance
 *   2) the `getRequestParams` function to build the request params.
 *
 * @param {Function} getRequestParams - A function that takes the `key` prop as an argument,
 *   and returns `[query, params]` which will be used by `queryFn` that actually
 *   calls the API.
 *
 * ### Optional props
 *
 * @param {Function} [afterFetchData] - A function to call after data has been fetched.
 *
 * @param {Function} [afterFetchNextPage] - A function to call after the next page has been fetched.
 */
```

Comment (RobertJoonas) [reply assets/js/dashboard/hooks/api-client.js]: Makes sense, thanks! https://github.com/plausible/analytics/pull/4361/commits/68bf8826e2072e7e6861181cb8fd0c98c0de675e
